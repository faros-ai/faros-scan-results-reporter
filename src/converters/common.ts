import { Mutation, QueryBuilder } from 'faros-js-client';
import { v4 } from 'uuid';

import { CodeQualityCategory, CodeQualityMetricType } from '../types';

export interface Config {
  readonly repoInfo?: {
    readonly name: string;
    readonly organization: string;
    readonly source: string;
  };
  readonly pullRequest?: number;
  readonly appInfo?: {
    readonly name: string;
    readonly platform: string;
  };
  readonly createdAt: string;
  readonly commit?: string;
  readonly branch?: string;
}

interface CodeCoverageMutationParams {
  coverageValue: number;
  commitSha?: string;
  createdAt: string;
  config: Config;
  qb: QueryBuilder;
}

export function createCodeCoveragePercentMutations(
  params: CodeCoverageMutationParams
): Array<Mutation> {
  const { coverageValue, commitSha, createdAt, config, qb } = params;
  const mutations = [];

  const qa_CodeQuality: any = {
    uid: v4(),
    coverage: {
      category: CodeQualityCategory.Coverage,
      type: CodeQualityMetricType.Percent,
      name: 'Coverage',
      value: coverageValue,
    },
    createdAt,
  };

  let repo;
  if (config.repoInfo) {
    repo = {
      name: config.repoInfo.name,
      organization: qb.ref({
        vcs_Organization: {
          uid: config.repoInfo.organization,
          source: config.repoInfo.source,
        },
      }),
    };
    qa_CodeQuality.repository = qb.ref({
      vcs_Repository: repo,
    });
    if (commitSha) {
      qa_CodeQuality.commit = qb.ref({
        vcs_Commit: {
          sha: commitSha,
          repository: qb.ref({
            vcs_Repository: repo,
          }),
        },
      });
    }
  }

  if (config.pullRequest && repo) {
    qa_CodeQuality.pullRequest = qb.ref({
      vcs_PullRequest: {
        number: config.pullRequest,
        repository: qb.ref({
          vcs_Repository: repo,
        }),
      },
    });
  }

  if (config.appInfo) {
    qa_CodeQuality.application = qb.ref({
      compute_Application: {
        name: config.appInfo.name,
        platform: config.appInfo.platform,
      },
    });
  }

  mutations.push(qb.upsert({ qa_CodeQuality }));
  if (config.branch && commitSha) {
    const branch = {
      name: config.branch,
      repository: qb.ref({
        vcs_Repository: repo,
      }),
    };

    const branchCommitAssociation = {
      commit: qb.ref({
        vcs_Commit: {
          sha: commitSha,
          repository: qb.ref({
            vcs_Repository: repo,
          }),
        },
      }),
      branch: qb.ref({
        vcs_Branch: branch,
      }),
    };

    mutations.push(qb.upsert({ vcs_Branch: branch }));
    mutations.push(qb.upsert({ vcs_BranchCommitAssociation: branchCommitAssociation }));
  }

  return mutations;
}
