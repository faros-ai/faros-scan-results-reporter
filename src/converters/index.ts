export * from './codeclimate';
export * from './common';
export * from './semgrep';
