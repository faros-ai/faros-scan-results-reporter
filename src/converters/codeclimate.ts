import { Mutation, QueryBuilder } from 'faros-js-client';

import { Converter } from '../converter';
import { Config, createCodeCoveragePercentMutations } from './common';

export interface CodeCoverageReport {
  git: {
    head: string;
    committed_at: number;
  };
  covered_percent: number;
}

export class CodeClimateConverter extends Converter {
  convert(
    data: CodeCoverageReport,
    config: Config,
    qb: QueryBuilder,
  ): Array<Mutation> {
    if (!data?.git?.head || !data?.covered_percent) {
      return [];
    }
    const coverageValue = data.covered_percent;
    const commitSha = data.git.head;
    const branch = config.branch;
    const createdAt = data?.git?.committed_at
      ? new Date(data.git.committed_at * 1000).toISOString()
      : config.createdAt;

    return createCodeCoveragePercentMutations({
      coverageValue,
      commitSha,
      createdAt,
      config,
      qb,
    });
  }
}
