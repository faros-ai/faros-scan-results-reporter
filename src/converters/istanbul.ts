import { Mutation, QueryBuilder } from 'faros-js-client';

import { Converter } from '../converter';
import { Config, createCodeCoveragePercentMutations } from './common';

export interface IstanbulCoverageReport {
  total: {
    lines: {
      pct: number;
    };
  };
}

export class IstanbulConverter extends Converter {
  convert(
    data: IstanbulCoverageReport,
    config: Config,
    qb: QueryBuilder,
  ): Array<Mutation> {
    if (!data?.total?.lines?.pct) {
      return [];
    }
    const coverageValue = data.total.lines.pct;
    const commitSha = config.commit;
    const createdAt = config.createdAt;

    return createCodeCoveragePercentMutations({
      coverageValue,
      commitSha,
      createdAt,
      config,
      qb,
    });
  }
}
