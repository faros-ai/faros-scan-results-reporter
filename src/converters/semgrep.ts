import { Mutation, QueryBuilder } from 'faros-js-client';
import { v4 } from 'uuid';

import { Converter } from '../converter';
import { CodeQualityCategory, CodeQualityMetricType } from '../types';
import { Config } from './common';

interface CodeLocation {
  readonly col: number;
  readonly line: number;
  readonly offset: number;
}

interface CodeSpan {
  readonly end: CodeLocation;
  readonly file: string;
  readonly start: CodeLocation;
}

interface ScanError {
  readonly code: number;
  readonly level: string;
  readonly message: string;
  readonly path: string;
  readonly spans: CodeSpan;
}

enum ScanResultCategory {
  Security = 'security',
}

interface ScanResultExtraMetadata {
  readonly category: string;
  readonly confidence: string;
  readonly cwe: Array<string>;
  readonly 'cwe2022-top25': boolean;
  readonly impact: string;
  readonly vulnerability_class: Array<string>;
}

interface ScanResultExtra {
  readonly engine_kind: string;
  readonly fingerprint: string;
  readonly is_ignored: boolean;
  readonly lines: string;
  readonly message: string;
  readonly metadata: ScanResultExtraMetadata;
  readonly severity: string;
}

interface ScanResult {
  readonly check_id: string;
  readonly start: CodeLocation;
  readonly end: CodeLocation;
  readonly extra: ScanResultExtra;
  readonly path: string;
}

interface SemgrepScanOutput {
  readonly errors: Array<ScanError>;
  readonly paths: {
    readonly scanned: Array<string>;
  };
  readonly results: Array<ScanResult>;
}

export class SemgrepConverter extends Converter {
  convert(
    data: SemgrepScanOutput,
    config: Config,
    qb: QueryBuilder,
  ): Array<Mutation> {
    const mutations = [];

    const qa_CodeQuality: any = {
      uid: v4(),
      vulnerabilities: {
        category: CodeQualityCategory.Security,
        type: CodeQualityMetricType.Int,
        name: 'vulnerabilities',
        value: '0',
      },
      createdAt: config.createdAt,
    };

    for (const result of data?.results ?? []) {
      if (result.extra?.metadata?.category === ScanResultCategory.Security) {
        qa_CodeQuality.vulnerabilities.value = addNtoString(
          qa_CodeQuality.vulnerabilities.value,
          1,
        );
      }
    }

    let repoInfo;
    if (config.repoInfo) {
      repoInfo = {
        name: config.repoInfo.name,
        organization: qb.ref({
          vcs_Organization: {
            uid: config.repoInfo.organization,
            source: config.repoInfo.source,
          },
        }),
      };
      qa_CodeQuality.repository = qb.ref({
        vcs_Repository: repoInfo,
      });
    }

    if (config.pullRequest && repoInfo) {
      qa_CodeQuality.pullRequest = qb.ref({
        vcs_PullRequest: {
          number: config.pullRequest,
          repository: qb.ref({
            vcs_Repository: repoInfo,
          }),
        },
      });
    }

    if (config.appInfo) {
      qa_CodeQuality.application = qb.ref({
        compute_Application: {
          name: config.appInfo.name,
          platform: config.appInfo.platform,
        },
      });
    }

    mutations.push(qb.upsert({ qa_CodeQuality }));

    return mutations;
  }
}

function addNtoString(numStr: string, n: number): string {
  return (Number(numStr) + n).toString();
}
