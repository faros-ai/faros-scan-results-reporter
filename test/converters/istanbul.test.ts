import { QueryBuilder } from 'faros-js-client';

import {
  IstanbulConverter,
  IstanbulCoverageReport,
} from '../../src/converters/istanbul';
import { Config } from '../../src/converters/common';

jest.mock('uuid', () => ({
  v4: jest.fn().mockReturnValue('my-uuid'),
}));

describe('IstanbulConverter', () => {
  let converter: IstanbulConverter;

  beforeEach(() => {
    converter = new IstanbulConverter();
  });

  it('should convert data to mutations', () => {
    const data: IstanbulCoverageReport = {
      total: {
        lines: {
          pct: 85,
        },
      },
    };

    const config: Config = {
      createdAt: '2024-03-28T16:20:05.474Z',
      repoInfo: {
        name: 'repo-name',
        organization: 'org-name',
        source: 'source-name',
      },
      pullRequest: 123,
      appInfo: {
        name: 'app-name',
        platform: 'platform-name',
      },
      commit: 'commit-hash',
      branch: 'main',
    };

    const result = converter.convert(
      data,
      config,
      new QueryBuilder('my-origin')
    );
    expect(result).toMatchSnapshot();
  });

  it('should return empty array for invalid data', () => {
    const invalidData = {};
    const config: Config = {
      createdAt: '2024-03-28T16:20:05.474Z',
    };

    const result = converter.convert(
      invalidData as IstanbulCoverageReport,
      config,
      new QueryBuilder('my-origin')
    );
    expect(result).toEqual([]);
  });
});
