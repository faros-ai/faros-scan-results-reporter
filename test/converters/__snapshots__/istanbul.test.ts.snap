// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IstanbulConverter should convert data to mutations 1`] = `
Array [
  Object {
    "mutation": Object {
      "insert_qa_CodeQuality_one": Object {
        "__args": Object {
          "object": Object {
            "application": Object {
              "data": Object {
                "name": "app-name",
                "platform": "platform-name",
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "compute_Application_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
            "commit": Object {
              "data": Object {
                "repository": Object {
                  "data": Object {
                    "name": "repo-name",
                    "organization": Object {
                      "data": Object {
                        "source": "source-name",
                        "uid": "org-name",
                      },
                      "on_conflict": Object {
                        "constraint": EnumType {
                          "value": "vcs_Organization_pkey",
                        },
                        "update_columns": Array [
                          EnumType {
                            "value": "refreshedAt",
                          },
                        ],
                      },
                    },
                  },
                  "on_conflict": Object {
                    "constraint": EnumType {
                      "value": "vcs_Repository_pkey",
                    },
                    "update_columns": Array [
                      EnumType {
                        "value": "refreshedAt",
                      },
                    ],
                  },
                },
                "sha": "commit-hash",
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "vcs_Commit_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
            "coverage": Object {
              "category": "Coverage",
              "name": "Coverage",
              "type": "Percent",
              "value": 85,
            },
            "createdAt": "2024-03-28T16:20:05.474Z",
            "origin": "my-origin",
            "pullRequest": Object {
              "data": Object {
                "number": 123,
                "repository": Object {
                  "data": Object {
                    "name": "repo-name",
                    "organization": Object {
                      "data": Object {
                        "source": "source-name",
                        "uid": "org-name",
                      },
                      "on_conflict": Object {
                        "constraint": EnumType {
                          "value": "vcs_Organization_pkey",
                        },
                        "update_columns": Array [
                          EnumType {
                            "value": "refreshedAt",
                          },
                        ],
                      },
                    },
                  },
                  "on_conflict": Object {
                    "constraint": EnumType {
                      "value": "vcs_Repository_pkey",
                    },
                    "update_columns": Array [
                      EnumType {
                        "value": "refreshedAt",
                      },
                    ],
                  },
                },
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "vcs_PullRequest_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
            "repository": Object {
              "data": Object {
                "name": "repo-name",
                "organization": Object {
                  "data": Object {
                    "source": "source-name",
                    "uid": "org-name",
                  },
                  "on_conflict": Object {
                    "constraint": EnumType {
                      "value": "vcs_Organization_pkey",
                    },
                    "update_columns": Array [
                      EnumType {
                        "value": "refreshedAt",
                      },
                    ],
                  },
                },
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "vcs_Repository_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
            "uid": "my-uuid",
          },
          "on_conflict": Object {
            "constraint": EnumType {
              "value": "qa_CodeQuality_pkey",
            },
            "update_columns": Array [
              EnumType {
                "value": "refreshedAt",
              },
              EnumType {
                "value": "uid",
              },
              EnumType {
                "value": "coverage",
              },
              EnumType {
                "value": "createdAt",
              },
              EnumType {
                "value": "repositoryId",
              },
              EnumType {
                "value": "commitId",
              },
              EnumType {
                "value": "pullRequestId",
              },
              EnumType {
                "value": "applicationId",
              },
              EnumType {
                "value": "origin",
              },
            ],
          },
        },
        "id": true,
      },
    },
  },
  Object {
    "mutation": Object {
      "insert_vcs_Branch_one": Object {
        "__args": Object {
          "object": Object {
            "name": "main",
            "origin": "my-origin",
            "repository": Object {
              "data": Object {
                "name": "repo-name",
                "organization": Object {
                  "data": Object {
                    "source": "source-name",
                    "uid": "org-name",
                  },
                  "on_conflict": Object {
                    "constraint": EnumType {
                      "value": "vcs_Organization_pkey",
                    },
                    "update_columns": Array [
                      EnumType {
                        "value": "refreshedAt",
                      },
                    ],
                  },
                },
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "vcs_Repository_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
          },
          "on_conflict": Object {
            "constraint": EnumType {
              "value": "vcs_Branch_pkey",
            },
            "update_columns": Array [
              EnumType {
                "value": "refreshedAt",
              },
              EnumType {
                "value": "name",
              },
              EnumType {
                "value": "repositoryId",
              },
              EnumType {
                "value": "origin",
              },
            ],
          },
        },
        "id": true,
      },
    },
  },
  Object {
    "mutation": Object {
      "insert_vcs_BranchCommitAssociation_one": Object {
        "__args": Object {
          "object": Object {
            "branch": Object {
              "data": Object {
                "name": "main",
                "repository": Object {
                  "data": Object {
                    "name": "repo-name",
                    "organization": Object {
                      "data": Object {
                        "source": "source-name",
                        "uid": "org-name",
                      },
                      "on_conflict": Object {
                        "constraint": EnumType {
                          "value": "vcs_Organization_pkey",
                        },
                        "update_columns": Array [
                          EnumType {
                            "value": "refreshedAt",
                          },
                        ],
                      },
                    },
                  },
                  "on_conflict": Object {
                    "constraint": EnumType {
                      "value": "vcs_Repository_pkey",
                    },
                    "update_columns": Array [
                      EnumType {
                        "value": "refreshedAt",
                      },
                    ],
                  },
                },
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "vcs_Branch_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
            "commit": Object {
              "data": Object {
                "repository": Object {
                  "data": Object {
                    "name": "repo-name",
                    "organization": Object {
                      "data": Object {
                        "source": "source-name",
                        "uid": "org-name",
                      },
                      "on_conflict": Object {
                        "constraint": EnumType {
                          "value": "vcs_Organization_pkey",
                        },
                        "update_columns": Array [
                          EnumType {
                            "value": "refreshedAt",
                          },
                        ],
                      },
                    },
                  },
                  "on_conflict": Object {
                    "constraint": EnumType {
                      "value": "vcs_Repository_pkey",
                    },
                    "update_columns": Array [
                      EnumType {
                        "value": "refreshedAt",
                      },
                    ],
                  },
                },
                "sha": "commit-hash",
              },
              "on_conflict": Object {
                "constraint": EnumType {
                  "value": "vcs_Commit_pkey",
                },
                "update_columns": Array [
                  EnumType {
                    "value": "refreshedAt",
                  },
                ],
              },
            },
            "origin": "my-origin",
          },
          "on_conflict": Object {
            "constraint": EnumType {
              "value": "vcs_BranchCommitAssociation_pkey",
            },
            "update_columns": Array [
              EnumType {
                "value": "refreshedAt",
              },
              EnumType {
                "value": "commitId",
              },
              EnumType {
                "value": "branchId",
              },
              EnumType {
                "value": "origin",
              },
            ],
          },
        },
        "id": true,
      },
    },
  },
]
`;
